### PR TITLE
Review

### PR DESCRIPTION
**Pull request by bzmo**

### CODE REVIEW

File: A4/RecursionToTheRescue.cpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        MinInfo minInfoWithCurrState = minPopularVoteToGetAtLeast(newVotesNeeded,
+                                                                  states,
+                                                                  minStateIndex + 1,
+                                                                  minInfocache);
+        minInfoWithCurrState.statesUsed.add(currState);
+        // Control integer overflow
+        if (minInfoWithCurrState.popularVotesNeeded < INT_MAX) {
+            minInfoWithCurrState.popularVotesNeeded += currState.popularVotes / 2 + 1;
+        }
+        // Cache the smaller of two results
```
Comment (bzmo) [inline A4/RecursionToTheRescue.cpp:286]: Ternary operator

File: A4/RecursionToTheRescue.cpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+ * the presidency.
+ *
+ * @param states All the states in the election (plus DC, if appropriate)
+ * @return Information about how few votes you'd need to win the election.
+ */
+MinInfo minPopularVoteToWin(const Vector<State>& states) {
+    MinInfo minVoteInfo;
+    minVoteInfo.popularVotesNeeded = 0;
+    minVoteInfo.statesUsed = {};
+    int totalElectoralVotes = sumElectoralVotesAboveIndex(states, 0);
```
Comment (bzmo) [inline A4/RecursionToTheRescue.cpp:310]: Use 'const' when variable assignments are not repeated

File: A4/RecursionToTheRescue.cpp [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    }
+    else if (unpreparedCities.isEmpty()) {
+        return true;
+    }
+    string city = unpreparedCities.first();
+    Set<string> tempUnpreparedCities = unpreparedCities;
+    addCityToStorageLocs(roadNetwork, city, locations, tempUnpreparedCities);
+    // Try current city
+    if (solveVertexCover(roadNetwork, numCities, locations, tempUnpreparedCities)) {
+        locations.add(city);
```
Comment (bzmo) [inline A4/RecursionToTheRescue.cpp:186]: `addCityToStorageLocs(...)` already calls `locations.add(city)`, so why add again?

File: A4/RecursionToTheRescue.cpp [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+}
+
+void addCityToStorageLocs(const Map<string, Set<string>>& roadNetwork,
+                          string city,
+                          Set<string>& locations,
+                          Set<string>& unpreparedCities) {
+    locations.add(city);
+    unpreparedCities.remove(city);
+    foreach(string neighbour in roadNetwork[city]) {
+        if (unpreparedCities.contains(neighbour)) {
```
Comment (bzmo) [inline A4/RecursionToTheRescue.cpp:223]: Double check how the set's `.remove(...)` handles non-existent elements because if it doesn't throw an exception, then it'd be cleaner to remove checking if `unpreparedCities.contains(...)`.

File: A4/RecursionToTheRescue.cpp [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+ */
+MinInfo minPopularVoteToWin(const Vector<State>& states) {
+    MinInfo minVoteInfo;
+    minVoteInfo.popularVotesNeeded = 0;
+    minVoteInfo.statesUsed = {};
+    int totalElectoralVotes = sumElectoralVotesAboveIndex(states, 0);
+    int electoralVotesNeeded = totalElectoralVotes/2 + 1;
+    SparseGrid<MinInfo> cacheMinInfo(totalElectoralVotes, totalElectoralVotes);
+    return minPopularVoteToGetAtLeast(electoralVotesNeeded, states, 0, cacheMinInfo);
+    return minVoteInfo;
```
Comment (bzmo) [inline A4/RecursionToTheRescue.cpp:314]: 2nd return statement?

File: A4/RecursionToTheRescue.cpp [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+    }
+    MinInfo currMinInfo;
+    // Base case
+    State currState = states[minStateIndex];
+    if (minStateIndex == states.size() - 1) {
+        if (electoralVotesNeeded > currState.electoralVotes) {
+            currMinInfo.popularVotesNeeded = INT_MAX;
+        }
+        else {
+            currMinInfo.popularVotesNeeded = currState.popularVotes / 2 + 1;
```
Comment (bzmo) [inline A4/RecursionToTheRescue.cpp:262]: NIT: I'd use a function here instead of repeating `votes/2 + 1` everywhere because it allows for easier maintainability and the function name provides some "documentation".

NIT is "Nitpicking" btw

File: A5/pqueue-vector.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    int minIndex = getMinIndex();
+    return pQueueVector[minIndex];
+}
+
+string VectorPQueue::extractMin() {
+    if (pQueueVector.isEmpty()) {
+        error("extractMin() : Attempting to extract from an empty queue.");
+    }
+    int minIndex = getMinIndex();
+    string minItem = peek();
```
Comment (bzmo) [inline A5/pqueue-vector.cpp:24]: Why call `peek()` after already calling `getMinIndex()`? Seems redundant

File: A5/pqueue.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+#include <typeinfo>
+using namespace std;
+
+/**
+ * Convenience function that converts a PQueue type
+ * to a string, so that information about PQueue type can
+ * be printed.
+ */
+string PQueue::typeToName(PQueueType type) {
+	switch (type) {
```
Comment (bzmo) [inline A5/pqueue.cpp:16]: The implementation here doesn't seem right to me because it goes against OOP, but if this is part of the sample code then w/e

File: A5/pqueue-heap.cpp [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+HeapPQueue::~HeapPQueue() {
+    delete [] array;
+}
+
+const string& HeapPQueue::peek() const {
+    if (logSize == 0) {
+        error("peek() : Attempting to peek at an empty queue.");
+    }
+    int root = 0;
+    return array[root];
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:19]: I'd just shorten this to array[0]

File: A5/pqueue-heap.cpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+void HeapPQueue::swapItems(int indexOne, int indexTwo) {
+    string tmp = array[indexOne];
+    array[indexOne] = array[indexTwo];
+    array[indexTwo] = tmp;
+}
+
+int HeapPQueue::findNextSwapIndex(int currIndex) {
+    int nextIndex = -1;
+    int lowerChildIndex = (currIndex + 1) * 2 - 1;
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:93]: Use 'const'

File: A5/pqueue-heap.cpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    int indexOfSmallerChild;
+    if (upperChildIndex >= logSize) { // Boundary case
+        if (array[currIndex] > array[lowerChildIndex]) {
+            return lowerChildIndex;
+        }
+    }
+    else { // General case
+        if (array[lowerChildIndex] < array[upperChildIndex]) {
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:106]: Use 'ternary' operator

File: A5/pqueue-heap.cpp [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+int HeapPQueue::findNextSwapIndex(int currIndex) {
+    int nextIndex = -1;
+    int lowerChildIndex = (currIndex + 1) * 2 - 1;
+    int upperChildIndex = lowerChildIndex + 1;
+    if (lowerChildIndex >= logSize) {
+        return nextIndex;
+    }
+
+    int indexOfSmallerChild;
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:99]: Move this declaration in the `else { ... } ` where it's needed

File: A5/pqueue-heap.cpp [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    array[indexOne] = array[indexTwo];
+    array[indexTwo] = tmp;
+}
+
+int HeapPQueue::findNextSwapIndex(int currIndex) {
+    int nextIndex = -1;
+    int lowerChildIndex = (currIndex + 1) * 2 - 1;
+    int upperChildIndex = lowerChildIndex + 1;
+    if (lowerChildIndex >= logSize) {
+        return nextIndex;
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:96]: Just return -1

File: A5/pqueue-heap.cpp [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+void HeapPQueue::swapItems(int indexOne, int indexTwo) {
+    string tmp = array[indexOne];
+    array[indexOne] = array[indexTwo];
+    array[indexTwo] = tmp;
+}
+
+int HeapPQueue::findNextSwapIndex(int currIndex) {
+    int nextIndex = -1;
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:92]: Change this to be `int invalidIndex = -1`

File: A5/pqueue-heap.cpp [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            indexOfSmallerChild = lowerChildIndex;
+        }
+        else {
+            indexOfSmallerChild = upperChildIndex;
+        }
+        if (array[indexOfSmallerChild] < array[currIndex]) {
+            nextIndex = indexOfSmallerChild;
+        }
+    }
+    return nextIndex;
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:116]: Return `invalidIndex`

File: A5/pqueue-heap.cpp [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    }
+    else { // General case
+        if (array[lowerChildIndex] < array[upperChildIndex]) {
+            indexOfSmallerChild = lowerChildIndex;
+        }
+        else {
+            indexOfSmallerChild = upperChildIndex;
+        }
+        if (array[indexOfSmallerChild] < array[currIndex]) {
+            nextIndex = indexOfSmallerChild;
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:113]: return `indexOfSmallerChild`

File: A5/pqueue-binomial-heap.cpp [hunk 17]
```diff
@@ -0,0 +1,158 @@
+#include "pqueue-binomial-heap.h"
```
Comment (bzmo) [inline A5/pqueue-binomial-heap.cpp:1]: TODO: Review this later because I'm lazy rn

File: A5/pqueue-heap.cpp [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+void HeapPQueue::enqueue(const string& elem) {
+    logSize ++;
+    if (logSize == capacity) {
+        expandCapacity();
+    }
+    int currIndex = logSize - 1;
+    int parentIndex = (currIndex + 1)/2 - 1;
+    array[currIndex] = elem;
+    while (parentIndex >= 0) {
```
Comment (bzmo) [inline A5/pqueue-heap.cpp:42]: Change this to be `while (parentIndex >= 0 && array[curIndex] < (array < parentIndex])) { ... } ` because the `else { break; }` looks unnecessary

File: A5/pqueue-linked-list.cpp [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+        }
+        else {
+            currVal = headTwo->data;
+            headTwo = headTwo->next;
+        }
+        currCell = new Cell(currVal);
+        insertCell(currCell, mergedTail, NULL);
+        mergedTail = mergedTail->next;
+    }
+    if (one->head != NULL) {
```
Comment (bzmo) [inline A5/pqueue-linked-list.cpp:75]: Ternary operator

File: A5/pqueue.cpp (hunk 8 above)
Comment (patrickz3li) [reply A5/pqueue.cpp:16]: yea it's part of sample code

File: A6/encoding.cpp [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+#include "strlib.h"
+#include "simpio.h"
+#include <fstream>
+#include <sstream>
+#include <bitset>
+using namespace std;
+
+Map<int, int> buildFrequencyTable(istream& input) {
+    Map<int, int> freqTable;
+    string line;
```
Comment (bzmo) [inline A6/encoding.cpp:16]: `string line` doesn't have a purpose

File: A6/encoding.cpp [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+    PriorityQueue<HuffmanNode *> pqNodes;
+    foreach (int asciiNum in freqTable) {
+        int frequency = freqTable[asciiNum];
+        HuffmanNode *node = new HuffmanNode(asciiNum, frequency);
+        pqNodes.enqueue(node, frequency);
+    }
+    while (pqNodes.size() > 1) {
+        HuffmanNode *nodeLeft = pqNodes.dequeue();
+        HuffmanNode *nodeRight = pqNodes.dequeue();
+        int freqSum = nodeLeft->count + nodeRight->count;
```
Comment (bzmo) [inline A6/encoding.cpp:35]: `const`

File: A6/encoding.cpp [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+        freqTable[asciiNum] ++;
+    }
+    freqTable[PSEUDO_EOF] ++; // add eof character
+    return freqTable;
+}
+
+HuffmanNode* buildEncodingTree(const Map<int, int>& freqTable) {
+    PriorityQueue<HuffmanNode *> pqNodes;
+    foreach (int asciiNum in freqTable) {
+        int frequency = freqTable[asciiNum];
```
Comment (bzmo) [inline A6/encoding.cpp:28]: `const`

File: A6/encoding.cpp [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+    return encodingMap;
+}
+
+void addToEncodingMap(HuffmanNode* encodingSubtree, Map<int, string>& encodingMap, string binaryPrefix) {
+    if (encodingSubtree == NULL) {
+        return;
+    }
+    addToEncodingMap(encodingSubtree->zero, encodingMap, binaryPrefix + '0');
+    addToEncodingMap(encodingSubtree->one, encodingMap, binaryPrefix + '1');
+    if (encodingSubtree->character != NOT_A_CHAR) {
```
Comment (bzmo) [inline A6/encoding.cpp:57]: I think you should move this check above the recursive calls because why recurse after you hit a leaf node?

File: A6/encoding.cpp [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+       encodingMap[encodingSubtree->character] = binaryPrefix;
+    }
+}
+
+void encodeData(istream& input, const Map<int, string>& encodingMap, obitstream& output) {
+    string binaryEncoding;
+    while(!input.eof()) {
+        int asciiNum = input.get();
```
Comment (bzmo) [inline A6/encoding.cpp:65]: `const`
